src/components/OAuth2Login/OAuth2Login.jsx

```diff
@@ -0,0 +1,156 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+
+import PopupWindow from './PopupWindow';
+import { toQuery } from './utils';
+
+const responseTypeLocationKeys = {
+  code: 'search',
+  token: 'hash',
+};
+
+const responseTypeDataKeys = {
+  code: 'code',
+  token: 'access_token',
+};
+
+class OAuth2Login extends Component {
+  constructor(props) {
+    super(props);
+    this.onBtnClick = this.onBtnClick.bind(this);
+    this.onRequest = this.onRequest.bind(this);
+    this.onSuccess = this.onSuccess.bind(this);
+    this.onFailure = this.onFailure.bind(this);
+  }
+
+  onBtnClick() {
+    const {
+      buttonText,
+      authorizationUrl,
+      clientId,
+      scope,
+      redirectUri,
+      state,
+      responseType,
+      popupWidth,
+      popupHeight,
+      isCrossOrigin,
+      extraParams,
+    } = this.props;
+    const payload = {
+      client_id: clientId,
+      scope,
+      redirect_uri: redirectUri,
+      response_type: responseType,
+      ...extraParams,
+    };
+    if (state) {
+      payload.state = state;
+    }
+    const search = toQuery(payload);
+    const width = popupWidth;
+    const height = popupHeight;
+    const left = window.screenX + ((window.outerWidth - width) / 2);
+    const top = window.screenY + ((window.outerHeight - height) / 2.5);
+    const locationKey = responseTypeLocationKeys[responseType];
+    const popup = PopupWindow.open(
+      buttonText,
+      `${authorizationUrl}?${search}`,
+      {
+        height, width, top, left,
+      },
+      {
+        locationKey,
+        isCrossOrigin,
+      },
+    );
+    this.popup = popup;
+
+    this.onRequest();
+    popup
+      .then(this.onSuccess)
+      .catch(this.onFailure);
+  }
+
+  onRequest() {
+    const { onRequest } = this.props;
+    onRequest();
```

Comment:
Could we switch this to be `this.props.onRequest();` rather than define a variable for it?